bpf/lib/nodeport.h

```diff
@@ -1617,12 +1617,16 @@ static __always_inline int encap_geneve_dsr_opt4(struct __ctx_buff *ctx, int l3_
 #endif
 
 	info = lookup_ip4_remote_endpoint(ip4->daddr, 0);
-	if (!info || info->tunnel_endpoint == 0)
+	if (!info)
 		return DROP_NO_TUNNEL_ENDPOINT;
 
-	tunnel_endpoint = info->tunnel_endpoint;
 	dst_sec_identity = info->sec_identity;
 
+	if (identity_is_remote_node(dst_sec_identity))
+		tunnel_endpoint = ip4->daddr;
```

Comment:
One unresolved question would be how this works for IPv6 (in `encap_geneve_dsr_opt6()`). We currently don't support overlay network over IPv6, so there we can't easily fall back to the backend IP :grimacing:.

The big hammer would be to touch https://github.com/cilium/cilium/blob/3d75eba2e41d6c0fd839feb4b4eb99904250edb3/pkg/node/manager/manager.go#L567. In native-routing that would nicely address this problem. But in overlay mode it would cause *all* `to-remote-node` traffic to flow via the overlay network. Which is something we've avoided so far ...